vernac/record.ml

```diff
@@ -592,20 +593,34 @@ let declare_structure ~cumulative finite ~univs ~variances ~primitive_proj
   let map i { Data.is_coercion; coers; rdata = { DataR.implfs; fields; _}; inhabitant_id; id; _ } =
     let rsp = (kn, i) in (* This is ind path of idstruc *)
     let cstr = (rsp, 1) in
-    let projections = declare_projections rsp (projunivs,ubinders) ~kind inhabitant_id coers implfs fields in
+    let projections = declare_projections rsp (projunivs,ubinders) ~kind:proj_kind inhabitant_id coers implfs fields in
     let build = GlobRef.ConstructRef cstr in
     let () = if is_coercion then ComCoercion.try_add_new_coercion build ~local:false ~poly in
     let struc = Structure.make (Global.env ()) rsp projections in
     let () = declare_structure_entry struc in
-    rsp
+    GlobRef.IndRef rsp
   in
-  List.mapi map record_data
+  List.mapi map record_data, []
 
 let implicits_of_context ctx =
   List.map (fun name -> CAst.make (Some (name,true)))
     (List.rev (Anonymous :: (List.map RelDecl.get_name ctx)))
 
-let build_class_constant ~univs ~rdata ~primitive_proj field implfs params paramimpls coers binder id proj_name =
+(* declare definitional class (typeclasses that are not record) *)
+let declare_class_constant ~univs paramimpls params data =
+  let id, rdata, coers, inhabitant_id = match data with
+    | [{ Data.id; rdata; is_coercion; coers; inhabitant_id }] ->
+       assert (not is_coercion);  (* should be ensured by caller *)
+       let coers = List.map (fun { pf_subclass; pf_priority } -> if pf_subclass then Some { hint_priority = pf_priority; hint_pattern = None } else None) coers in
+       id, rdata, coers, inhabitant_id
+    | _ -> assert false in  (* should be ensured by caller *)
```

Comment:
Not a fan of this, why are we taking a list in the first place?

Reply:
Agree, this code is duplicated and should go in a specific function. The list comes from the fact that record (unlike typeclasses) can be mutually recursive records (with the `Inductive ... := { ... } with ... := { ... }` syntax.